host_modules/image_service.py

```diff
@@ -0,0 +1,135 @@
+        try:
+            response = requests.get(image_url, stream=True)
+            if response.status_code != 200:
+                logger.error(
+                    "Failed to download image: HTTP status code {}".format(
+                        response.status_code
+                    )
+                )
+                return errno.EIO, "HTTP error: {}".format(response.status_code)
+
+            with tempfile.NamedTemporaryFile(dir="/tmp", delete=False) as tmp_file:
+                for chunk in response.iter_content(chunk_size=8192):
+                    tmp_file.write(chunk)
+                temp_file_path = tmp_file.name
+            os.rename(temp_file_path, save_as)
+            return 0, "Download successful"
+        except Exception as e:
+            logger.error("Failed to write downloaded image to disk: {}".format(e))
+            return errno.EIO, str(e)
+
+    @host_service.method(
+        host_service.bus_name(MOD_NAME), in_signature="s", out_signature="is"
+    )
+    def install(self, where):
+        """
+        Install a a sonic image:
+
+        Args:
+            where: either a local path or a remote url pointing to the image.
+        """
+        logger.info("Using sonic-installer to install the image at {}.".format(where))
+        cmd = ["/usr/local/bin/sonic-installer", "install", "-y", where]
+        result = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        msg = ""
+        if result.returncode:
+            lines = result.stderr.decode().split("\n")
+            for line in lines:
+                if "Error" in line:
+                    msg = line
+                    break
+        return result.returncode, msg
+
+    @host_service.method(
+        host_service.bus_name(MOD_NAME), in_signature="ss", out_signature="is"
+    )
+    def checksum(self, file_path, algorithm):
+        """
+        Calculate the checksum of a file.
+
+        Args:
+            file_path: path to the file.
+            algorithm: checksum algorithm to use (sha256, sha512, md5).
+        """
+
+        logger.info("Calculating {} checksum for file {}".format(algorithm, file_path))
+
+        if not os.path.isfile(file_path):
+            logger.error("File {} does not exist".format(file_path))
+            return errno.ENOENT, "File does not exist"
+
+        hash_func = None
+        if algorithm == "sha256":
+            hash_func = hashlib.sha256()
+        elif algorithm == "sha512":
+            hash_func = hashlib.sha512()
+        elif algorithm == "md5":
+            hash_func = hashlib.md5()
+        else:
+            logger.error("Unsupported algorithm: {}".format(algorithm))
+            return errno.EINVAL, "Unsupported algorithm"
+
+        try:
+            with open(file_path, "rb") as f:
+                for chunk in iter(lambda: f.read(4096), b""):
+                    hash_func.update(chunk)
+            return 0, hash_func.hexdigest()
```

Comment:
Can we use md5sum directly? Is this hash result the same as md5sum result? How long will it take to calculate checksum for a typical sonic image?

Reply:
1-2 seconds. md5sum is slightly faster but not by a noticeable margin. And since this is python code, I think we should use python library to have better control over the logic.
```
$ python3
Python 3.11.2 (main, May  2 2024, 11:59:08) [GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from md5 import calculate_md5
>>> calculate_md5('sonic.bin')
(2.589256763458252, '489a131693a4c98803d5a520df409bf9')
>>>
$ time md5sum sonic.bin
489a131693a4c98803d5a520df409bf9  sonic.bin

real    0m1.879s
user    0m1.621s
sys     0m0.257s

```